PostgreSQL 10 converts unknown type to text type

yahonda opened the pull request:
### Summary

Address #28799

* Refer these discussion and commit:
https://www.postgresql.org/message-id/28163.1486478547%40sss.pgh.pa.us
https://git.postgresql.org/gitweb/?p=postgresql.git&a=commitdiff&h=1e7c4bb0049732ece651d993d03bb6772e5d281a

* Skip 2 specs which expect `select_all` returns unknown type if PostgreSQL 10 or higher
There was a suggestion to remove these warning spec but this behavior is true
as long as PostgreSQL 9.6 or lower version, then just skip them.

* Add 1 spec to validate unknown type is converted to text type if PostgreSQl 10 or higher

### Other Information
This pull request has been tested with these two versions of PostgreSQL.

```sql
activerecord_unittest=# select version();
                                                  version
-----------------------------------------------------------------------------------------------------------
 PostgreSQL 10devel on x86_64-pc-linux-gnu, compiled by gcc (GCC) 6.3.1 20161221 (Red Hat 6.3.1-1), 64-bit
(1 row)
```

```
activerecord_unittest=# select version();
                                                 version
----------------------------------------------------------------------------------------------------------
 PostgreSQL 9.5.6 on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 6.2.0-5ubuntu12) 6.2.0 20161005, 64-bit
(1 row)

activerecord_unittest=#
```

matthewd commented:
I think the use of "unknown" in these test names is a bit misleading: their intention is to test behaviour for *unrecognised* type oids -- not `unknown` specifically.

We probably just need to find a more version-compatible choice of type to use, which won't get converted to `text`... possibly by having them define a custom type for that purpose.

yahonda commented:
Let me confirm if I understand your comment correctly.

These `test_only_reload_type_map_once_for_every_unknown_type` and `test_only_warn_on_first_encounter_of_unknown_oid`  are testing how Rails handles any unrecognized type which is not recognized(registered) by `initialize_type_map` method. 

One of these not recognized types is PostgreSQL unknown type whose OID is 705. 
I do not know any other unrecognized type though; I guess it could happen if PostgreSQL introduces new OID and Rails does not handle it.

In PostgreSQL 10, testing unrecognized type using OID may not be appropriate since OID 705 will be "known" by Rails since PostgreSQL 10 converts unknown OID type result into Text. 

Then you are suggesting to create another type which is not recognized by Rails yet and use this new type at these two tests.

yahonda committed:
PostgreSQL 10 converts unknown OID 705 to text 25

- Rename test cases from  `unknown` to `unrecognized` since unknown OID
is one of possible unrecognized types by Rails

- Use "select 'pg_catalog.pg_class'::regclass" whose OID is 2205, which
will not be converted to recognized type in PostgreSQL 10.

activerecord_unittest=# select oid, typname from pg_type where oid in (2205, 2277);
 oid  | typname
------+----------
 2205 | regclass
 2277 | anyarray
(2 rows)

Addresses #28868

matthewd commented:
Exactly.

We do encounter unknown types -- possibly because of future PostgreSQL versions, but more likely due to extensions, or types defined by the user. The challenge is just in coming up with one that's predictable enough for our test suite, yet not common enough to include in our default type map. :smile:

We register `oid`, but `regclass` & friends might come through as unknown. Or `xid`?

I think I'd start by trying `'pg_catalog'::regnamespace`: it avoids any meta-ambiguity that might come from using `regtype`, say, but also references something that's guaranteed to always exist.

yahonda commented:
Updated the pull request to use "select 'pg_catalog.pg_class'::regclass" whose OID is 2205, which
will not be converted to recognized type in PostgreSQL 10.

I do not have PostgreSQL 9.1 environment though, I have confirmed this `pg_class` exists at PostgreSQL 9.1 which is the minimum requirement for current Rails master branch.

matthewd commented:
👌🏻 perfect, thank you!

> I do not have PostgreSQL 9.1 environment though, I have confirmed this pg_class exists at PostgreSQL 9.1 which is the minimum requirement for current Rails master branch.

Yeah, `pg_class` has been there forever. The `regclass` etc types are newer, but still 8.x-era.
